### PR TITLE
GUACAMOLE-518: Correct Caps Lock and Shift behavior within RDP keymap translation system.

### PR DESCRIPTION
**Pull request by mike-jumper**
This change restructures the RDP keymap translation system such that:

* Overall modifier and lock state are tracked, not just the state of individual keysyms.
* Multiple possible key definitions may be associated with each keysym. For example, the `en-us-qwerty` keymap describes both `Caps+Shift+A` and `A` as possible ways to type a lowercase "a".
* When a key is pressed, the key definition which requires the least complex additional events will be chosen based on the current keyboard lock/modifier state.
* The PDU sent from the RDP server when the lock state changes remotely is handled, keeping local tracking of keyboard state in sync.
* Rather than release automatically-pressed keys immediately after pressing the required scancode, maintain consistent state, instead releasing _all_ automatically-pressed keys only after the user has released all manually-pressed keys client-side.

With this in place, modifiers and locks are set only as required, and the behavior resulting in [GUACAMOLE-518](https://issues.apache.org/jira/browse/GUACAMOLE-518) no longer occurs.

This unfortunately required painstakingly researching and editing all currently-supported keymaps ... with the exception of the Japanese keymap which has additional issues and will need to be corrected at a later date via [GUACAMOLE_520](https://issues.apache.org/jira/browse/GUACAMOLE-520). The results of that research can be found here:

https://issues.apache.org/jira/browse/GUACAMOLE-518?focusedCommentId=17142479&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17142479

### CODE REVIEW

File: src/protocols/rdp/keyboard.c [hunk 1]
```diff
@@ -248,37 +446,21 @@ static guac_rdp_key* guac_rdp_keyboard_get_key(guac_rdp_keyboard* keyboard,
  *     The keymap to use to populate the given client's keysym/scancode
  *     mapping.
  */
-static void __guac_rdp_keyboard_load_keymap(guac_rdp_keyboard* keyboard,
+static void guac_rdp_keyboard_load_keymap(guac_rdp_keyboard* keyboard,
```
Comment (necouchman) [inline src/protocols/rdp/keyboard.c:449]: If this is still `static`, and thus presumably only used within the context of this file, why the change in name?

File: src/protocols/rdp/Makefile.am [hunk 2]
```diff
@@ -241,7 +241,7 @@ rdp_keymaps =                                \
     $(srcdir)/keymaps/da_dk_qwerty.keymap    \
     $(srcdir)/keymaps/tr_tr_qwerty.keymap
 
-_generated_keymaps.c: $(rdp_keymaps)
+_generated_keymaps.c: $(rdp_keymaps) $(srcdir)/keymaps/generate.pl
 	$(AM_V_GEN) $(srcdir)/keymaps/generate.pl $(rdp_keymaps)
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply src/protocols/rdp/Makefile.am:245]: Yep. The overall syntax of `Makefile` rules is:

```
output: list of files required for producing output
    command to produce output
```

In this case, `generate.pl` should be listed as it is part of the source tree and changes to the script require a rebuild of the output. Until added here, changes that I made to `generate.pl` during testing required manually deleting `_generated_keymaps.c` to force a rebuild.

File: src/protocols/rdp/keyboard.c (hunk 1 above)
Comment (mike-jumper) [reply src/protocols/rdp/keyboard.c:449]: Apparently, the double underscore prefix is actually a reserved namespace in C. It's supposed to be dedicated to the use of the compiler. Having learned this, I've been switching use of `__guac_*` to `guac_*` wherever I encounter it.

File: src/protocols/rdp/keyboard.c [hunk 3]
```diff
@@ -217,20 +315,120 @@ static guac_rdp_key* guac_rdp_keyboard_map_key(guac_rdp_keyboard* keyboard,
...
+/**
+ * Given a key which may have multiple possible definitions, returns the
+ * definition that currently has the lowest cost, taking into account the
+ * current keyboard lock and modifier states.
+ *
+ * @param keyboard
+ *     The guac_rdp_keyboard associated with the current RDP session.
+ *
+ * @param key
+ *     The key being whose lowest-cost possible definition should be retrieved.
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply src/protocols/rdp/keyboard.c:343]: Yikes, me too. I'll reword.

[the remark this answers is not in the document]
Comment (mike-jumper) [reply src/protocols/rdp/keyboard.c]: OK - I've removed that "being" that snuck in there somehow.
